### PR TITLE
All true players sleeping Fixed!

### PR DESCRIPTION
**Pull request by Sanluli36li**
completely rewritten
#136 gnembon/fabric-carpet#425
Fix the problem that can only be reset to day 1
Now it will not check the player's sleep every tick More efficient
All behave more like Vanilla

thx gnembon's suggest

### CODE REVIEW

**Comment by Sanluli36li** [comment]
Now only executed in overworld, no longer check other dimensions

**Comment by Sanluli36li** [comment]
Known issues: Ignore doDaylightCycle and doWeatherCycle rules, Because they can cause the player to not get up, I don’t know yet there is any way in Scarpet to get players up.

File: programs/survival/sleep_ignore_fake_players.sc [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+// player wakes up
+__on_player_wakes_up(player) ->
+(
+    __update_sleeping_players();
+);
+
+// player interacts with bed
+__on_player_interacts_with_block(player, hand, block, face, hitvec) ->
+(
+    for([
```
Comment (gnembon) [inline programs/survival/sleep_ignore_fake_players.sc:46]: you can just check if the block ends with '_bed' using `block ~ '_bed$'`

File: programs/survival/sleep_ignore_fake_players.sc [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        );
+        global_allPlayerSleep = j > 0 && j >= length(entity_list('player')) - i;
+    );
+);
+
+__check_sleeping_enough() ->
+(
+    for(entity_list('player'),
+        if(_~'gamemode' == 'spectator' || _~'player_type' == 'fake' || query(_, 'nbt', 'SleepTimer') >= 100,
+            continue(),
```
Comment (gnembon) [inline programs/survival/sleep_ignore_fake_players.sc:79]: continue can just be replaced with 'null' (as a noop) and return at the end, just with `true`

File: programs/survival/sleep_ignore_fake_players.sc [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+__config() -> { 'scope' -> 'global', 'stay_loaded' -> true };
+
+global_allPlayerSleep = false;
+
+__on_tick() ->
+(
+    in_dimension('overworld',
+        if(global_allPlayerSleep && __check_sleeping_enough(),
+            global_allPlayerSleep = false;
+            if(run('gamerule doDaylightCycle'),
```
Comment (gnembon) [inline programs/survival/sleep_ignore_fake_players.sc:16]: run() now returns a triple (three results) so you would need to adjust to that.

**Comment by Ghoulboy78** [comment]
Add name to list of contributors

**Pull request merged**